src/mame/video/alpha68k_palette.cpp

```diff
@@ -133,34 +160,91 @@ uint16_t alpha68k_palette_device::read(offs_t offset)
 	return m_paletteram[offset];
 }
 
-inline void alpha68k_palette_device::set_color_entry(u16 offset, u16 pal_data, int shift)
+inline void alpha68k_palette_device::set_color_entry(u16 offset)
 {
+	const u16 pal_data = m_paletteram[offset];
+
 	int dark = pal_data >> 15;
 	int r = ((pal_data >> 14) & 0x1) | ((pal_data >> 7) & 0x1e);
 	int g = ((pal_data >> 13) & 0x1) | ((pal_data >> 3) & 0x1e);
 	int b = ((pal_data >> 12) & 0x1) | ((pal_data << 1) & 0x1e);
 
-	r >>= shift;
-	g >>= shift;
-	b >>= shift;
+	r >>= m_sync_color_shift[0];
+	g >>= m_sync_color_shift[0];
+	b >>= m_sync_color_shift[0];
 
 	set_pen_color(offset, m_palette_lookup[r][dark], m_palette_lookup[g][dark], m_palette_lookup[b][dark]);
+	// TODO: Shadow is used?
 }
 
 void alpha68k_palette_device::write(offs_t offset, u16 data, u16 mem_mask)
 {
 	COMBINE_DATA(&m_paletteram[offset]);
-	u16 pal_data = m_paletteram[offset];
-	set_color_entry(offset, pal_data, 0);
+	update_color(offset, offset);
+}
 
+inline void alpha68k_palette_device::update_color(u16 entry, u16 offset)
+{
 	// reference color
-	if (offset == 0)
+	if (entry == 0)
 	{
 		// TODO: actual behaviour, needs HW tests.
-		bool is_sync_color = (pal_data & 0x7fff) == 0;
+		bool is_sync_color = (m_paletteram[offset] & 0x7fff) == 0;
 		int sync_color_shift = is_sync_color ? 0 : 2;
+		if (m_sync_color_shift[0] != sync_color_shift)
+		{
+			m_sync_color_shift[0] = sync_color_shift;
+			for (int i=1; i<m_entries; i++)
+				set_color_entry(i);
+		}
+	}
+	set_color_entry(offset);
+}
+
+uint16_t neogeo_palette_device::read(offs_t offset)
+{
+	return m_paletteram[m_bankaddr + offset];
+}
+
+inline void neogeo_palette_device::set_color_entry(u16 offset)
+{
+	const u16 pal_data = m_paletteram[offset];
```

Comment:
I'm not convinced over writing then reading the very same palette data performance-wise, unless you have a profiler output that proves otherwise.